core/cloudflow-spark-testkit/src/main/scala/cloudflow/spark/testkit/SparkStreamletTestkit.scala

```diff
@@ -256,3 +257,55 @@ case class SparkOutletTap[T: Encoder](
   // get results from memory sink
   def asCollection(session: SparkSession): Seq[T] = session.sql(s"select * from $queryName").as[T].collect()
 }
+case class ExecutionReport(totalRows: Long, totalQueries: Int, failures: Seq[String]) {
+  def hasFailures = failures.nonEmpty
+}
+
+class QueryExecutionMonitor()(implicit ec: ExecutionContext) extends StreamingQueryListener {
+  @volatile var status: Map[UUID, QueryState] = Map()
+  @volatile var dataRows: Map[UUID, Long]     = Map()
+
+  sealed trait QueryState
+  case object Started extends QueryState
+  case class Terminated(exception: Option[String]) extends QueryState {
+    def isFailed = exception.nonEmpty
+  }
+
+  def executionReport: ExecutionReport = {
+    val exceptions = status.values.collect { case Terminated(Some(reason)) => reason }
+    ExecutionReport(dataRows.values.sum, dataRows.keys.size, exceptions.toSeq)
+  }
+
+  def waitForData(): Future[Long] = {
+    def hasData   = dataRows.nonEmpty && dataRows.forall { case (_, rows) => rows > 0 }
+    def totalRows = dataRows.values.sum
+    Future {
+      while (!hasData) {
+        Thread.sleep(1.second.toMillis)
+      }
+      totalRows
```

Comment:
How do we ensure here that we have all the rows that need to be fetched ? Should `waitForData` take an argument `expected_no_of_rows` so that it can match ?

Reply:
In testkit we always will be dealing with a finite data set - right ?

Reply:
I don't think we can say it is always finite. It depends on how the end user uses it.

Reply:
We have existing test cases using infinite data sources (`rate` source)